Add relative and absolute tolerances for matrix_rank, pinv

IvanYashchuk opened the pull request:
This pull request introduces new keyword arguments for `torch.linalg.matrix_rank` and `torch.linalg.pinv`: `atol` and `rtol`.

Currently, only tensor overload has default values for either `atol` or `rtol`, the float overload requires both arguments to be specified.

FC compatibility: https://github.com/pytorch/pytorch/pull/63102#discussion_r710930509

Fixes https://github.com/pytorch/pytorch/issues/54151. Fixes https://github.com/pytorch/pytorch/issues/66618.


cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

IvanYashchuk committed:
Add atol, rtol arguments for matrix_rank

IvanYashchuk committed:
Add atol, rtol for pinv

IvanYashchuk committed:
Add a test for matrix_rank atol rtol interaction

IvanYashchuk committed:
Do not change the original overload name, it's bc breaking

IvanYashchuk committed:
Fix Caffe2 build failures because of ambiguous declarations

IvanYashchuk committed:
Wrapper function should be after _add_docstr

IvanYashchuk committed:
Remove python wrapper, it breaks jit

IvanYashchuk committed:
Update docs with current defaults

IvanYashchuk committed:
flake8

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -923,13 +923,15 @@
 The matrix rank is computed as the number of singular values
 (or eigenvalues in absolute value when :attr:`hermitian`\ `= True`)
-that are greater than the specified :attr:`tol` threshold.
+that are greater than `max(atol, σ₁*rtol)` threshold,
```
We settled for not using unicode because of rendering issues. This should be in LaTeX (see also the next line)

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -923,13 +923,15 @@
 
 matrix_rank = _add_docstr(_linalg.linalg_matrix_rank, r"""
-matrix_rank(A, tol=None, hermitian=False, *, out=None) -> Tensor
+linalg.matrix_rank(A, *, atol=None, rtol=None, hermitian=False, out=None) -> Tensor
+linalg.matrix_rank(A, tol, hermitian=False, *, out=None) -> Tensor
```
How is this rendered in the docs? How do we want to advertise this? Perhaps just leave the new one but secretly support the old one? cc @mruberry

IvanYashchuk replied:
There is actually an error when trying to build the docs using the current version. Probably there should be an empty line.
This double function signature was inspired by [Julia docs for pinv](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.pinv).
I think we should leave the new signature as main one and in the description tell that there is also NumPy compatible version.

Lezcano replied:
Completely agree. I'd say we go with that.

Lezcano replied:
Also, about the Numpy compat, see https://github.com/pytorch/pytorch/pull/63102#discussion_r689158004

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -939,17 +941,15 @@
 .. math::
 
-    \text{tol} = \sigma_1 \max(m, n) \varepsilon
+    \text{rtol} = \max(m, n) \varepsilon
```
This might be better inline now that's not too complicated.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1654,7 +1660,19 @@
 The singular values (or the norm of the eigenvalues when :attr:`hermitian`\ `= True`)
-that are below the specified :attr:`rcond` threshold are treated as zero and discarded in the computation.
+that are below `max(atol, σ₁*rtol)` threshold are treated as zero and discarded in the computation,
+where :math:`\sigma_1` is the largest singular value (or eigenvalue).
```
Same points as for `matrix_rank` in this and the next lines.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -939,17 +941,15 @@
-:math:`\varepsilon` is the epsilon value for the dtype of :attr:`A` (see :class:`torch.finfo`).
-If :attr:`A` is a batch of matrices, :attr:`tol` is computed this way for every element of
+and :math:`\varepsilon` is the epsilon value for the dtype of :attr:`A` (see :class:`torch.finfo`).
+If :attr:`A` is a batch of matrices, :attr:`rtol` is computed this way for every element of
```
Even though it was like this in the previous version, I believe it's safe to delete it here and in the docs of `pinv`, as this is the expected behaviour really.

Lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10461,11 +10482,27 @@
   python_module: linalg
 
-- func: linalg_matrix_rank(Tensor self, float? tol=None, bool hermitian=False) -> Tensor
+- func: linalg_matrix_rank.atol_rtol_tensor(Tensor input, *, Tensor? atol=None, Tensor? rtol=None, bool hermitian=False) -> Tensor
```
Nice that these args are kwonly now.

Lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10394,15 +10394,36 @@
+  python_module: linalg
+  variants: function
+
+# It's not possible to specify both atol and rtol as optional with floats that would work in C++
+# because calling linalg_pinv(self) would be ambiguous, it clashes with the tensor overload.
+# Having one of the float as optional would conflict with the old rcond overload
+# Therefore for floats both values need to be specified and there is no default.
```
Make this into a note and cite it in `matrix_rank` as well.

IvanYashchuk replied:
I removed this note completely. The problem was fixed with `cpp_no_default_args`.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -408,39 +504,43 @@ static Tensor& linalg_matrix_rank_out_helper(const Tensor& input, const Tensor&
-
-  result = linalg_matrix_rank_out_helper(input, atol, rtol, hermitian, result);
+Tensor& linalg_matrix_rank_out(const Tensor& input, double tol, bool hermitian, Tensor& result) {
+  // For NumPy compatibility tol is not scaled with max(singular_value) if the value for tol is provided
+  // It is assumed that the provided value is the absolute tolerance
```
Is this documented? It's a pretty annoying behaviour really...

IvanYashchuk replied:
In the current PR there is a sentence: "If the value of :attr:`tol` is `None` then it's equivalent to setting `rtol=tol` and `atol=tol` otherwise."

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -231,18 +277,61 @@ Tensor linalg_pinv(const Tensor& input, const Tensor& rcond, bool hermitian) {
 Tensor linalg_pinv(const Tensor& input, double rcond, bool hermitian) {
-  Tensor rcond_tensor = at::full({}, rcond, input.options().dtype(ScalarType::Double));
-  return at::linalg_pinv(input, rcond_tensor, hermitian);
+  // For NumPy compatibility the rcond argument is used as relative tolerance
```
Same as for `matrix_rank`, is this documented?

IvanYashchuk replied:
In this PR I added a sentence "If the value of :attr:`rcond` is specified then it's equivalent to setting `rtol=rcond`."

Lezcano replied:
We discussed with @mruberry the option of start having a section of "compatibility arguments" after the "Keyword arg:" section in the docs. This section would advertise the arguments saying "this is just here for compatibility with XXX framework". Perhaps this could be a fantastic start for this section? Add something like:
> Compatibility args:
>   rcond (float, Tensor, optional): [Numpy] alias for ``:attr:`rtol` ``.

IvanYashchuk replied:
I added the Compatibility args section.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -231,18 +277,61 @@ Tensor linalg_pinv(const Tensor& input, const Tensor& rcond, bool hermitian) {
+
+Tensor linalg_pinv(const Tensor& input, const Tensor& rcond, bool hermitian) {
+  // For NumPy compatibility the rcond argument is used as relative tolerance
+  TORCH_CHECK(!at::isComplexType(rcond.scalar_type()),
+              "torch.linalg.pinv: rcond tensor of complex type is not supported.");
```
Could you factor out all these errors into `LinearAlgebraUtils.h` for consistency and easy of use? Same for `atol` and `acond` and so on.

IvanYashchuk replied:
Something like `checkNotComplex`? But this error check occurs in the code only in this function.

Lezcano replied:
It's for these two functions and also `lstsq` whenever we support it, right?

IvanYashchuk replied:
Done.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -192,18 +192,63 @@ std::tuple<Tensor, Tensor> slogdet(const Tensor& self) {
+// This function extracts the optional Tensors for atol and rtol
+// Default value for atol is zero
+// Default value for rtol is eps*max(rows, cols)
+// If atol is specified and rtol is not specified then default value for rtol is zero
```
I believe this is not documented. Could you add it?

IvanYashchuk replied:
Added a sentence "If :attr:`rtol` is not specified and :attr:`atol` is specified to be larger than zero then
:attr:`rtol` is set to zero."

lezcano reviewed (commented):
Left a number of comments.

As per the offline discussion, I think that it would be interesting to consider running a number of tests with the proposed defaults of `rtol = 5*(sqrt(n) + sqrt(m))*eps`. As the linear algebra expert we consulted noted, the bounds `rtol = max(m, n)*eps` are based on classical worst-case error analysis, which are known (and shown) to be pessimistic. (sic)

I think we have a chance to improve on this. Furthermore, the deviation from the current behaviour would not be too large, as this function is expected to have to be tuned to work for the given matrix. As such, I believe it'd be beneficial to change the default behaviour of this function for it to work in more cases, so it doesn't bite people.

We have the numpy scripts  that they used when they decided this value:
http://www.mail-archive.com/numpy-discussion@scipy.org/msg37819.html
and we could also run the script that gives wrong results with the current defaults provided by @nikitaved in https://github.com/pytorch/pytorch/issues/54151#issuecomment-897568523.

There's also the talk that this professor recommended: https://www.youtube.com/watch?v=B1eFGn5nN84 It'd be good to have a look at it before making a decision on this topic.

IvanYashchuk committed:
Use cpp_no_default_args to fix C++ ambiguity

IvanYashchuk committed:
Add a note that rtol is set to zero by default if atol is specified

IvanYashchuk committed:
Update docs

Remove unicode
Update latex to inline instead of separate block
Remove old function signature
Add a note that there is a NumPy compatible variant

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into rtol-atol-fix

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into rtol-atol-fix

IvanYashchuk committed:
Add common checkNotComplexTolerance function

IvanYashchuk committed:
Add comaptibility args section

Add a note on broadcasting of tolerances arguments

IvanYashchuk commented:
Thanks, @Lezcano, for the review! I think I addressed all the comments. The only thing left is whether we want to change the default relative tolerance in this PR. I'm inclined to use what propagates from Matlab (`max(m, n)` * eps`) and introduce the change separately if we decide to do so.

IvanYashchuk committed:
Compatibility args section does not render correctly, remove it

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1684,15 +1705,17 @@
-                                       Default: `1e-15`.
-    hermitian(bool, optional): indicates whether :attr:`A` is Hermitian if complex
-                               or symmetric if real. Default: `False`.
+    rcond (float, Tensor, optional): [NumPy compatibility] the tolerance value.
+                                     If the value of :attr:`rcond` is specified then
+                                     it's equivalent to setting `rtol=rcond`.
```
What about the simpler:
```suggestion
    rcond (float, Tensor, optional): [NumPy Compat]. Alias for :attr:`rtol` . Default: `None`.
```

Also, should we put this here (the place where it semantically belongs), or at the end of `Keyword args`?

I am more inclined for the latter. This way people could find it in the docs, and we would still (silently) support NumPy programs, but we would announce this kind of args as little as possible. cc @mruberry

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1654,7 +1660,18 @@
 
 The singular values (or the norm of the eigenvalues when :attr:`hermitian`\ `= True`)
-that are below the specified :attr:`rcond` threshold are treated as zero and discarded in the computation.
+that are below :math:`\max(\text{atol}, \sigma_1 * \text{rtol})` threshold are
```
Nit.
```suggestion
that are below :math:`\max(\text{atol}, \sigma_1 \cdot \text{rtol})` threshold are
```

IvanYashchuk committed:
Try fixing finfo link

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -961,12 +962,17 @@
-                                   Default: `None`.
-    hermitian(bool, optional): indicates whether :attr:`A` is Hermitian if complex
-                               or symmetric if real. Default: `False`.
+    tol (float, Tensor, optional): [NumPy compatibility] the tolerance value.
+                                   If the value of :attr:`tol` is `None`
+                                   then it's equivalent to setting `rtol=tol` and `atol=tol` otherwise.
```
What about the simpler:
```suggestion
    tol (float, Tensor, optional): [NumPy Compat] Alias for :attr:`atol`. Default: `None`.
```
If it's specified to `None` or not specified then this function does what it normally does by default, right?

IvanYashchuk replied:
Right, I applied your suggestion.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1654,7 +1660,18 @@
-that are below the specified :attr:`rcond` threshold are treated as zero and discarded in the computation.
+that are below :math:`\max(\text{atol}, \sigma_1 * \text{rtol})` threshold are
+treated as zero and discarded in the computation,
+where :math:`\sigma_1` is the largest singular value (or eigenvalue).
```
Is this line correct? If so, we should use the squared norm of the eigenvalue of largest norm, to be consistent between the two definitions.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -231,18 +286,60 @@ Tensor linalg_pinv(const Tensor& input, const Tensor& rcond, bool hermitian) {
     Tensor S_abs = S.abs();
```
It should be squared in this case, as the singular values are the squared norm of the eigenvalues.

IvanYashchuk replied:
The code is correct, the singular values are the absolute values of the eigenvalues of a Hermitian matrix.

Lezcano replied:
Ah true, you are completely right.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1669,6 +1686,10 @@
 
+.. note::
+    This function has NumPy compatible variant `linalg.pinv(A, rcond, hermitian=False)`.
+    However, use of the positional argument :attr:`rcond` is deprecated in favor of :attr:`rtol`.
```
Perhaps not "deprecated" but... but what? what could we say here? @mruberry

lezcano reviewed (commented):
Left a number of comments. Many of them are about how to go with these sort of weird compatibility args. We should decide how to treat them here and be consistent in the future. @mruberry will surely have some ideas in mind on how to go about this.

IvanYashchuk committed:
Update torch/linalg/__init__.py

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk committed:
Apply suggestions from code review

Use "[NumPy Compat] Alias for ..."

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk commented:
CI failures are related to this PR and I need to fix it.

IvanYashchuk committed:
Change tol_value in tests to make the result deterministic

IvanYashchuk committed:
Add matrix_rank and pinv to allowlist for bc check

IvanYashchuk committed:
Use a different function for testing lite interpreter

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into rtol-atol-fix

IvanYashchuk committed:
Try avoiding singular error in DefaultArgsTensorinvSpecifyDefault

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10425,15 +10425,33 @@
   variants: function
 
-- func: linalg_pinv(Tensor self, float rcond=1e-15, bool hermitian=False) -> Tensor
+- func: linalg_pinv.atol_rtol_tensor(Tensor self, *, Tensor? atol=None, Tensor? rtol=None, bool hermitian=False) -> Tensor
```
This is cool but changing the signatures like this is FC-breaking because a torch program that just calls linalg.pinv on a tensor will now serialize this call, and if that program is exported to an older version of PyTorch it won't understand this signature.

We need to add these signatures for 2+ weeks before switching to them as the defaults. (2 weeks is our forward-compatibility window.)

This also wouldn't matter if on one was using linalg.pinv, but it already has adoption, which is cool.

IvanYashchuk replied:
Alright, I will investigate what actually happens in this case maybe it doesn't break anything.

mruberry replied:
Let's try serializing these on this build and loading them in PyTorch 1.9 using torchscript and see if they'll still run when `linalg.pinv.(t)` is called.

mruberry replied:
(An alternative may be to make a Python wrapper that dispatches `linalg.pinv(t)` and `linalg.pinv(t, *args, **kwargs)` separately.)

IvanYashchuk replied:
Summary:

- `torch.linalg.pinv(x)` _is not_ FC-breaking,
- `torch.linalg.pinv(x, rcond: float)` _is not_ FC-breaking,
- `torch.linalg.pinv(x, hermitian: bool)` _**is**_ FC-breaking.

Honestly, it's still a mystery to me now how this works, but it does and here's how I verified it:

On current PR save .pt file using `torch.jit.script`
```
In [1]: import torch

In [2]: @torch.jit.script
   ...: def test(x):
   ...:     return torch.linalg.pinv(x)
   ...:

In [3]: test.graph
Out[3]:
graph(%x.1 : Tensor):
  %4 : bool = prim::Constant[value=0]()
  %2 : NoneType = prim::Constant()
  %5 : Tensor = aten::linalg_pinv(%x.1, %2, %2, %4) # <ipython-input-2-609569be27ef>:3:11
  return (%5)

In [4]: test.save("test_pinv.pt")
```
Load on other branch ([for example 1.9 in colab](https://colab.research.google.com/drive/1LBIxKZsqcwAcwFiiiIwNTHvAK6FM8h9J?usp=sharing)):
```
In [1]: import torch

In [2]: test = torch.jit.load("test_pinv.pt")

In [3]: test.graph
Out[3]:
graph(%self : __torch__.PlaceholderModule,
      %x.1 : Tensor):
  %6 : bool = prim::Constant[value=0]()
  %5 : float = prim::Constant[value=1.0000000000000001e-15]()
  %7 : Tensor = aten::linalg_pinv(%x.1, %5, %6) # <ipython-input-2-609569be27ef>:3:11
  return (%7)

In [4]: test(torch.randn(3, 3))
Out[4]:
tensor([[ 0.1149,  0.0695, -0.5795],
        [-0.4950, -0.3870, -0.1382],
        [-0.1383,  0.4591,  0.3753]])
```
The interesting thing is that `test.graph` is different. I do not understand yet how `torch.linalg.pinv(x)` and `torch.linalg.pinv(x, hermitian: bool)` are different. @mruberry, do we need to fix `torch.linalg.pinv(x, hermitian: bool)`?

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10492,11 +10510,29 @@
 
-- func: linalg_matrix_rank(Tensor self, float? tol=None, bool hermitian=False) -> Tensor
+- func: linalg_matrix_rank.atol_rtol_tensor(Tensor input, *, Tensor? atol=None, Tensor? rtol=None, bool hermitian=False) -> Tensor
+  python_module: linalg
```
See above FC concern.

This may require offline discussion for how to resolve. I've added it to our linalg meeting schedule.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into rtol-atol-fix

IvanYashchuk committed:
Fix matrix_rank signature in autocast_mode.cpp

mruberry reviewed (approved):
I think we're OK on BC/FC. Would you just rebase this, @IvanYashchuk?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into rtol-atol-fix

IvanYashchuk commented:
> I think we're OK on BC/FC. Would you just rebase this, @IvanYashchuk?

This is great!
I rebased, but now there is a clash of signatures because `linalg_pinv` has an explicit derivative rule since  https://github.com/pytorch/pytorch/pull/66092.

IvanYashchuk committed:
Update derivatives.yaml signature for linalg_pinv

IvanYashchuk committed:
Fix derivative signature

IvanYashchuk committed:
Add tensor rcond for OpInfo sample inputs of pinv

IvanYashchuk committed:
Revert changes to sample_inputs_linalg_pinv_hermitian
